hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/execution/OMGateway.java

```diff
@@ -0,0 +1,226 @@
+  private final LeaderRequestExecutor leaderExecutor;
+  private final OzoneManager om;
+  private final OmLockOpr omLockOpr;
+  private final AtomicLong requestInProgress = new AtomicLong(0);
+  private final AtomicLong uniqueIndex = new AtomicLong();
+
+  public OMGateway(OzoneManager om) throws IOException {
+    this.om = om;
+    this.omLockOpr = new OmLockOpr(om.getThreadNamePrefix());
+    this.leaderExecutor = new LeaderRequestExecutor(om, uniqueIndex);
+    if (om.isRatisEnabled()) {
+      om.getOmRatisServer().getOmStateMachine().getPostExecutor().registerIndexNotifier(this::indexNotifier);
+    } else {
+      long index = PostExecutor.initLeaderIndex(om);
+      uniqueIndex.set(index);
+    }
+  }
+
+  public void start() {
+    Iterator<Map.Entry<CacheKey<String>, CacheValue<OmBucketInfo>>> itr
+             = om.getMetadataManager().getBucketTable().cacheIterator();
+    while (itr.hasNext()) {
+      Map.Entry<CacheKey<String>, CacheValue<OmBucketInfo>> entry = itr.next();
+      if (entry.getValue().getCacheValue() != null) {
+        QuotaResource.Factory.registerQuotaResource(entry.getValue().getCacheValue().getObjectID());
+      }
+    }
+    omLockOpr.start();
+    leaderExecutor.start();
+  }
+  public void stop() {
+    leaderExecutor.setEnabled(false);
+    leaderExecutor.stop();
+    omLockOpr.stop();
+  }
+
+  public OMResponse submit(OMRequest omRequest, byte[] clientId, long callId) throws ServiceException {
+    // TODO handle replay of request for external request
+    return submitInternal(omRequest, getClientId(clientId), getCallId(callId));
+  }
+
+  public OMResponse submit(OMRequest omRequest) throws ServiceException {
+    return submitInternal(omRequest, null, 0L);
+  }
+
+  public OMResponse submitInternal(OMRequest omRequest, String clientId, long callId) throws ServiceException {
+    requestInProgress.incrementAndGet();
+    RequestContext requestContext = new RequestContext();
+    requestContext.setRequest(omRequest);
+    requestContext.setFuture(new CompletableFuture<>());
+    CompletableFuture<OMResponse> f = requestContext.getFuture().whenComplete(
+        (r, th) -> handleAfterExecution(requestContext, th));
+    OmLockOpr.OmLockInfo lockInfo = null;
+    try {
+      OMRequestExecutor requestExecutor = OmRequestFactory.createRequestExecutor(omRequest, om);
+      OMRequest request = requestExecutor.preProcess(om);
+      // re-update modified request from pre-process
+      requestContext.setRequest(request);
+      requestContext.setUuidClientId(clientId);
+      requestContext.setCallId(callId);
+      requestContext.setRequestExecutor(requestExecutor);
+
+      OzoneManagerRequestHandler.requestParamValidation(omRequest);
+
+      requestContext.getRequestExecutor().authorize(om);
+      lockInfo = requestContext.getRequestExecutor().lock(omLockOpr);
+
+      validatePrepareState(requestContext.getRequest());
+      ensurePreviousRequestCompletionForPrepare(requestContext.getRequest());
```

Comment:
There is no guarantee that Previous Requests will go down to zero? For each request, you will need to check if there is prepare pending.